src/main/java/com/patreon/euphrates/ScrubbingInputStream.java

```diff
@@ -6,90 +6,88 @@
+    byteBuffer = ByteBuffer.allocate(4096);
+    charBuffer = CharBuffer.allocate(4096);
+    goodCharBuffer = CharBuffer.allocate(4096);
+    byteBuffer.limit(0);
   }
 
   @Override
   public int read() throws IOException {
-    int b = in.read();
-    return charValid(b) ? b : read();
+    fillBuffer();
+
+    if (!byteBuffer.hasRemaining()) return -1;
+    int b = (int)byteBuffer.get();
+    return b;
   }
 
   @Override
   public int read(byte[] b, int off, int len) throws IOException {
-    int r = in.read(b, off, len);
-    if (r > 0) {
-      int lastChar = off;
-      // scan all chars
-      for (int pos = off; pos < off + r; pos++) {
-        // if invalid
-        if (!charValid(b[pos])) {
-          // find the next valid one
-          for (int nextValid = pos + 1; nextValid < off + r; nextValid++) {
-            // and swapem!
-            if (charValid(b[nextValid])) {
-              byte cc = b[nextValid];
-              b[nextValid] = b[pos];
-              b[pos] = cc;
-              lastChar = pos;
-              break;
-            }
-          }
-        } else {
-          lastChar = pos;
-        }
-      }
-      r += lastChar - (off + r - 1);
+    fillBuffer();
+    if (!byteBuffer.hasRemaining()) {
+      return -1;
     }
-    return r;
+
+    int maxAvailableLength = Math.min(byteBuffer.remaining(), len);
+    byteBuffer.get(b, off, maxAvailableLength);
+    return maxAvailableLength;
   }
 
-  private boolean charValid(int c) {
-    switch (c) {
-      // these characters are forbidden by xml 1.0 which is what mysqldump produces,
-      // however, stream itself is not xml 1.0 compliant, thus, remove control characters
-      // except for the printable ones
-      case 0x0:
-      case 0x1:
-      case 0x2:
-      case 0x3:
-      case 0x4:
-      case 0x5:
-      case 0x6:
-      case 0x7:
-      case 0x8:
-        //case 0x9: tab      \t
-        //case 0xa: new line \n
-      case 0xb:
-      case 0xc:
-        // case 0xd: cr      \r
-      case 0xe:
-      case 0xf:
-      case 0x10:
-      case 0x11:
-      case 0x12:
-      case 0x13:
-      case 0x14:
-      case 0x15:
-      case 0x16:
-      case 0x17:
-      case 0x18:
-      case 0x19:
-      case 0x1a:
-      case 0x1b:
-      case 0x1c:
-      case 0x1d:
-      case 0x1e:
-      case 0x1f:
-        return false;
-      default:
-        return true;
+  private void fillBuffer() throws IOException {
+    if (byteBuffer.hasRemaining()) {
+      return;
+    }
+
+    int bytesRead = in.read(readBuffer);
+    boolean eof = bytesRead == -1;
+
+    ByteBuffer bytesIn = ByteBuffer.wrap(readBuffer, 0, eof ? 0 : bytesRead);
+    decoder.decode(bytesIn, charBuffer, eof);
+    if (eof) decoder.flush(charBuffer);
+
+    charBuffer.flip();
+
+    while (charBuffer.hasRemaining()) {
+      char c = charBuffer.get();
+      if (isPrintableChar(c)) {
+        goodCharBuffer.put(c);
+      }
     }
+    charBuffer.compact();
```

Comment:
Is this ever going to do anything but get to the end of the buffer?  You're reading to eof of `bytesIn`, reading until charBuffer.hasRemaining is `false`...then you're compacting, which means you're at the end of the buffer; I think maybe doing a `clear` at the beginning of `fillBuffer` would be clearer and have the same effect.